[그래] 로또 2단계 미션 제출합니다

hwanghe159 opened the pull request:
안녕하세요! 두가지 질문이 있습니다!! 😄 

## 첫번째 질문
2단계가 되면서 수동 로또 개수에 대한 요구사항이 생기다 보니까
수동 로또 개수, 자동 로또 개수 둘 다 관리하는 `LottoCount`클래스를 만들었습니다.
그리고 `LottoCount`클래스 내부에는 수동로또 개수인 `manualCount`와 자동로또 개수인`automaticCount`가 있습니다.
그런데 제가 구현한 생성자를 보시면     
`public LottoCount(int manualCount, int automaticCount)` 가 아니라
`public LottoCount(int totalCount, int manualCount)` 입니다.

main에서 먼저 
전체 총 로또개수를 먼저 구하게 되고, 그 다음 수동로또 개수를 입력받고,
전체로또개수와 수동로또개수를 이용하여 자동로또개수를 구하기 때문에 이렇게 구현했습니다.
그리고 이렇게 하지 않으면 main에서 (전체로또개수 - 수동로또개수)를 하는 과정이 추가로 들어가기 때문에
조금 더 깔끔하게 보이고 싶은 마음에 이렇게 구현했습니다.

위와 같이 개발자의 편의?를 위해서 임의로 생성자의 인자를 배치해도 되는건지,
아니면 `public LottoCount(int manualCount, int automaticCount)` 이렇게
내부 변수를 차례로 인자로 갖는 것이 권장사항인건지 궁금합니다.




## 두번째 질문
제가 전략 패턴을 사용해서 코드를 개선하기 위해서 전략패턴을 공부중입니다.
그런데 직접 구현하려고 보니 한 가지 문제가 있었습니다.

`LottoCreationStrategy` 인터페이스를 생성해서 `createLottos(int count)` 추상 메서드를 배치하고,
`AutomaticCreationStrategy`와 `ManualCreationStrategy`가 `LottoCreationStrategy` 인터페이스를 구현하도록 할 예정이었습니다.
>`AutomaticCreationStrategy`의 `createLottos(int count)`는 자동로또 개수를 인자로 받아서 그 개수만큼 생성하여 반환,
`ManualCreationStrategy`의 `createLottos(int count)`는 수동로또 개수를 인자로 받아서 그 개수만큼 생성하여 반환

`AutomaticCreationStrategy`의 `createLottos(int count)`는 구현에 문제가 없었습니다. 
왜냐하면 자동 로또의 개수(count)만 인자로 전달받으면 알아서 만들어서 반환하면 되니까요.
그런데 `ManualCreationStrategy`의 `createLottos(int count)`를 구현하다 보니 의문점이 생겼습니다.
수동구매는 count만큼 구매할 번호를 반복적으로 입력을 받아야 하는데 
이렇게 구현하면 `createLottos(int count)`가 InputView를 의존하게 되는 형태가 되었습니다.

전에 Domain이 View에 의존하면 안된다는 피드백을 받아왔어서 위와 같이 구현하면 안된다는 생각에
전략 패턴을 사용했던 다른 크루의 코드를 보니 Strategy안에 View를 사용하는 코드가 머지된 걸 봤습니다.

혹시 위처럼 구현하는 게 허용되는건지, 아니면 안되는건지
그리고 더 좋은 방법이 있는건지 궁금합니다.


-------------
리뷰 감사합니다!! 😄

Commit:
docs : README 작성

Commit:
docs : README 수정

Commit:
feat : 로또 구입 금액을 입력받고 검증 후 예외처리하는 PurchasePrice 클래스 생성

Commit:
feat : 로또 구입 개수를 계산하는 메서드 추가, 출력하는 메서드 추가

Commit:
feat : 로또 한 줄을 생성하는 LottoGenerator 클래스 생성

Commit:
feat : 로또 한 줄의 정보를 담는 Lotto 클래스 생성

Commit:
feat : 로또 한 줄의 숫자를 출력하는 메서드를 뷰에 추가

Commit:
feat : 구입한 모든 로또를 관리하는 Lottos 클래스 생성

Commit:
feat : 구입한 모든 로또의 번호를 출력하는 printLottos() 생성

Commit:
docs : README 수정

당첨 번호와 보너스 번호에 대한 예외 추가 정리

Commit:
feat : 당첨 번호를 입력받아 검증하는 WinningNumbers 클래스 생성, 테스트코드 작성

Commit:
feat : 보너스 번호를 검증하는 BonusNumber 클래스 생성

Commit:
docs : README 수정

Commit:
feat : 당첨 결과를 계산하는 로직 추가

Lotto클래스에 로또 한 줄과 당첨 번호를 비교해서 비교 결과를 반환하는 메서드 추가
Lottos클래스에 로또 여러줄과 당첨 번호를 비교해서 전체 로또 결과를 반환하는 메서드 추가

Commit:
feat : 당첨 결과 계산을 위한 메서드 추가, 결과를 나타내는 MatchResult enum 생성

Commit:
test : 당첨 결과 반환 로직 테스트

Commit:
feat : 당첨 결과를 관리하고 수익률을 계산하는 LottoResult 클래스 생성

Commit:
test : 수익률에 관한 테스트 추가

Commit:
feat : 해당되는 당첨 결과가 없을 때 처리 로직 추가

Commit:
feat : 로또 당첨 결과와 수익률 출력 기능 추가

Commit:
refactor : 메서드 분리와 순서, 상수 정리

Commit:
test : 테스트 코드 상수명 변경과 필요없는 공백 삭제

Commit:
docs : README 수정

Commit:
refactor : 로또 번호를 나타내는 Ball 클래스 생성에 따른 리팩토링

로또 번호와 관련된 모든 Integer 을 Ball로 감쌈
그에따른 재명명(BonusNumber -> BonusBall, WinningNumbers -> WinningBalls)
각 로또 번호에 대한 검증 로직을 모두 Ball 클래스 내부로 이전

Commit:
refactor : 리팩토링

1~45의 모든 Ball을 생성하는 로직을 Ball클래스로 옮겨 여러 클래스에 퍼져 있는 중복 상수(MIN_NUMBER, MAX_NUMBER 등) 통합
모든 당첨 결과를 나타내는 LottoResult를 Results로 재명명

Commit:
Merge branch 'hwanghe159' into step2

Commit:
refactor : 피드백 반영 (Arrays.asList 사용하여 리팩토링)

Commit:
refactor : WinningLotto클래스로 WinningBalls와 BonusBall 통합

Commit:
refactor : 로또 번호와 당첨 번호를 비교하여 결과를 도출하는 로직을 Lotto와 Results클래스 내부로 이동

Commit:
refactor : 정적 팩토리 메서드를 사용하여 당첨 결과 반환, Lotto클래스에서 중복을 검증하는 코드 추가, 리팩토링에 따른 테스트 코드 수정

Commit:
feat : 수동로또 개수와 자동로또 개수를 관리하는 LottoCount클래스 생성

Commit:
feat : 수동로또 기능이 추가됨에 따른 View 기능 추가

Commit:
feat : 수동 로또 개수만큼 반복적으로 입력받는 로직 추가

Commit:
feat : 잘못 입력했을때 throw되는 InvalidInputException 추가

Commit:
refactor : 예외 종류 변경, 메서드 재명명

jeonghoon1107 commented on src/main/java/lotto/LottoApplication.java:
```diff
@@ -4,18 +4,29 @@
+
+        OutputView.requestManualLottoMessage();
+        Lottos lottos = new Lottos();
+        for (int i = 0; i < lottoCount.getManualCount(); i++) {
```
LottoGenerator에는 자동 로또만 받아오게 되어있는데 interface로 추출해보면 어떨까요?
간단하게 예를 들어보면
interface는 `List<Lotto> generate()` 메서드를 갖고 내부 구현은 감추어도 좋을 것 같아요. 필요하다면 추가로 파라미터를 받아도 좋고.
LottoGeneratorFactory가 AutoLottoGenerator, ManualLottoGenerator를 가지고 Lottos를 반환하는 식으로.
단순한 예시니 참고만 하시면 좋을 것 같아요!

jeonghoon1107 reviewed (changes requested):
안녕하세요 그래!
구조에 관련한 피드백을 남겼으니 확인해주세요 :)
추가적인 질문이 있으면 DM 주세요!

1. 생성자 인자의 위치
보통은 변수의 순서대로 인자를 만들어주는 것이 혼동이 덜 할 것 같아요. (IDE에서 생성자를 만들 때도 순서를 지켜서 만들어주기도 하고요)
이렇게 타입이 같은 상황에서는 특히 혼동의 여지가 더 클 것 같은데요, 저 같은 경우에는 빌더를 이용하는 편입니다!

2. Domain과 View는 최대한 분리하는 것이 좋습니다.
View의 변경으로 인해 비즈니스 코드에 영향이 가는 것은 최대한 줄이는 것이 좋을 것 같다는 생각입니다!

hwanghe159 committed:
refactor : 컨트롤러 생성, main으로부터 분리

hwanghe159 committed:
refactor : 수동 로또 생성하는 부분 분리, LottoCount의 변수 변경

hwanghe159 committed:
refactor : LottoCreationStrategy 인터페이스 생성

hwanghe159 committed:
refactor : 모든 Ball을 다루는 캐시를 Ball클래스로 이동, LottoCreationStrategy의 모든 구현클래스 구현 완료

hwanghe159 committed:
refactor : 필요없는 주석 제거, 컨벤션에 맞게 메서드 위치 이동

hwanghe159 committed:
refactor

6개의 당첨 번호를 관리하는 WinningBalls클래스와 보너스 볼을 관리하는 BonusBall클래스 제거,
WinningBalls를 Lotto클래스와 통합, BonusBall을 Ball클래스와 통합하여 중복 상당수 제거,
WinningLotto를 WinningNumbers로 재명명,
생성자 오버로딩으로 중복으로 검증하는 로직 제거,
메서드 이름을 좀 더 명확하게 재명명, 메서드 순서 재배치

hwanghe159 committed:
refactor : 필요없어 보이는 메서드 제거, 스트림 이용하여 리팩토링

hwanghe159 committed:
refactor : 테스트 코드 컴파일 오류 해결

hwanghe159 committed:
refactor : 필요없는 주석 제거

jeonghoon1107 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,39 @@
+        PurchasePrice purchasePrice = new PurchasePrice(InputView.requestPurchasePrice());
+        LottoCount lottoCount =
+                new LottoCount(purchasePrice.calculateTotalLottoCount(), InputView.requestNumberOfManualLotto());
+        Lottos lottos = createManualLottos(lottoCount.getManualCount());
```
lottoCount를 넘겨서 manual과 auto lotto를 한번에 만들어 주는 별도의 클래스를 구현하면 어떨까요?

jeonghoon1107 commented on src/main/java/lotto/domain/strategy/ManualCreationStrategy.java:
```diff
@@ -0,0 +1,29 @@
+
+public class ManualCreationStrategy implements LottoCreationStrategy {
+
+    private List<List<String>> manualLottos;
```
List<String>으로도 충분한 것 아닌가요?

jeonghoon1107 commented on src/main/java/lotto/domain/strategy/AutomaticCreationStrategy.java:
```diff
@@ -0,0 +1,34 @@
+        return new Lottos(lottos);
+    }
+
+    public Lotto createOneLotto() {
```
접근제어자를 private으로 변경하면 어떨까요?

jeonghoon1107 commented on src/main/java/lotto/domain/strategy/AutomaticCreationStrategy.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Override
+    public Lottos create() {
```
dto를 넘기는 방식으로 구현하면 깔끔할 것 같아요.
```
public interface LottoCreationStrategy {
     Lottos create(dto)
}
```

jeonghoon1107 reviewed (approved):
안녕하세요 그래!
리팩토링 잘 하셨네요.
간단한 피드백을 남겼으니 확인해주세요.
해당 pr은 머지할게요!
궁금한 점이 있으면 DM 주세요.
